### PR TITLE
ansible-code2.12 + ansible-runner updates

### PR DESCRIPTION
**Pull request by dangel101**
- Move ansible-runner runtime dir to /var/lib
- Require ansible-core and ansible-runner
- ansible runner for debug
- removing private key
- ansible-runner playbook execution
- ansible-core2.12 + ansible-runner

### CODE REVIEW

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleCommandConfig.java [hunk 1]
```diff
@@ -212,4 +258,102 @@ public AnsibleCommandConfig logFileSuffix(String logFileSuffix) {
...
+        }
+        ansibleCommand.add(ANSIBLE_COMMAND);
+        ansibleCommand.add(ANSIBLE_EXECUTION_METHOD);
+        ansibleCommand.add(AnsibleConstants.HOST_DEPLOY_PROJECT_DIR);
+        ansibleCommand.add("-p");
+        ansibleCommand.add(playbook);
+        ansibleCommand.add("--artifact-dir");
+        ansibleCommand.add(AnsibleConstants.ARTIFACTS_DIR);
+        ansibleCommand.add("--inventory");
+        ansibleCommand.add(String.valueOf(inventoryFile));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleCommandConfig.java:281]: Wouldn't it be better to use `inventoryFile.toAbsolutePath()` instead?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleCommandConfig.java [hunk 2]
```diff
@@ -212,4 +258,102 @@ public AnsibleCommandConfig logFileSuffix(String logFileSuffix) {
...
+        ansibleCommand.add("-i");
+        ansibleCommand.add(String.valueOf(uuid));
+        return ansibleCommand;
+    }
+
+    private void createInventoryFile() {
+        Path inventoryFile;
+        if (getInventoryFile() == null) {
+//             If hostnames are empty we just don't pass any inventory file:
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleCommandConfig.java:290]: When could this happen?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java [hunk 3]
```diff
@@ -50,241 +37,276 @@
...
-        }
-    }
-
-    public String runPlaybook(AnsibleCommandConfig command) {
-        if (command.hosts() != null) {
-            for (VDS host : command.hosts()) {
-                addHost(host.getHostName(), host.getSshPort());
+    public AnsibleReturnValue artifactHandler(UUID uuid, int lastEventID, int timeout, BiConsumer<String, String> fn)
+            throws Exception { // TODO is this the correct place?
+        log.error(String.format("*** inside artifacts handle*** %1$s", Thread.currentThread().getName()));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java:71]: Maybe mode to debug level?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java [hunk 4]
```diff
@@ -50,241 +37,276 @@
...
-            Comparator.comparing(s -> Integer.parseInt(s.substring(0, s.indexOf("-"))))
-        );
-        while (it.hasNext()) {
-            String n = it.next();
-            int index = Integer.parseInt(n.substring(0, n.indexOf("-")));
-            if (index > lastEventId) {
-                set.add(n);
+    public void runPlaybook(List<String> command, int timeout) throws Exception {
+        final Object executeLock = new Object();
+        log.error(String.format("***inside run playbook*** %1%s", Thread.currentThread()));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java:237]: Maybe mode to debug level?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java [hunk 5]
```diff
@@ -50,241 +37,276 @@
...
+        Process ansibleProcess;
+        File output = File.createTempFile("output", ".log");
+        synchronized (executeLock) {
+            log.error(String.format("***inside run playbook, syncronized block*** %1$s %2$s", Thread.currentThread().getName(), command.get(command.size() -1)));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java:241]: Maybe mode to debug level?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java [hunk 6]
```diff
@@ -50,241 +37,276 @@
...
+            ProcessBuilder ansibleProcessBuilder =
+                    new ProcessBuilder(command).redirectErrorStream(true).redirectOutput(output);
+            String str = String.join(" ", command);
+            log.error(String.format("command: %1$s", str));
+            ansibleProcess = ansibleProcessBuilder.start();
+            log.error(String.format("***inside run playbook*** after starting the process %1$s %2$s", Thread.currentThread().getName(), command.get(command.size() -1)));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java:247]: Maybe mode to debug level?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java [hunk 7]
```diff
@@ -50,241 +37,276 @@
...
+            if (!ansibleProcess.waitFor(timeout, TimeUnit.MINUTES)) {
+                throw new Exception("Timeout occurred while executing Ansible playbook.");
+            }
+            if (ansibleProcess.exitValue() != 0) {
+                throw new AnsibleRunnerCallException(
+                        "Failed to execute call to start playbook. %1$s",
+                        output.toString()); // TODO: need to pass not path but content
             }
         }
+        log.error(String.format("***inside run playbook*** finished running the process. %1$s %2$s", Thread.currentThread().getName(), command.get(command.size() -1)));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java:257]: Maybe mode to debug level?

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java [hunk 8]
```diff
@@ -50,241 +37,276 @@
...
+        JsonNode currentNode = null;
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            String jsonOutput = Files.readString(Paths.get(eventPath), StandardCharsets.UTF_8);
+            if (!jsonIsValid(jsonOutput)) {
+//                throw new AnsibleRunnerCallException(
+//                        "Failed to fetch info about event. %1$s: %2$s",
+//                        RunnerJsonNode.status(eventDir), //TODO check where and how to retrieve status & msg?
+//                        RunnerJsonNode.msg(eventDir)
+//                );
```
Comment (mwperina) [inline backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClient.java:303]: Why is above code commented out? If not needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 9]
```diff
@@ -52,190 +34,190 @@
...
+//    }
+
+//    private static HttpResponse createHttpResponse(String content, int httpStatus) {
+//        HttpResponseFactory factory = new DefaultHttpResponseFactory();
+//        HttpResponse response = factory.newHttpResponse(
+//            new BasicStatusLine(HttpVersion.HTTP_1_1, httpStatus, null), null
+//        );
+//        response.setEntity(EntityBuilder.create().setText(content).build());
+//        return response;
+//    }
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:65]: Why is the above code commented out? If non needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 10]
```diff
@@ -52,190 +34,190 @@
...
+//        when(httpClient.execute(any(HttpPost.class))).thenReturn(
+//            createHttpResponse(
+//                "{\"status\": \"FAILED\", \"msg\": \"Runner thread failed to start\"}",
+//                HttpStatus.SC_INTERNAL_SERVER_ERROR
+//            )
+//        );
+
+//        assertThrows(AnsibleRunnerCallException.class, () -> {
+//            client.runPlaybook(command);
+//        });
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:96]: Why is the above code commented out? If non needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 11]
```diff
@@ -52,190 +34,190 @@
...
-        HttpResponse response
-    ) throws IOException {
-        when(httpClient.execute(any(HttpGet.class))).thenReturn(response);
+//    @ParameterizedTest(name = "Test playbook statuses: {0}")
+//    @MethodSource("statusesForPlaybook")
+//    public void getPlaybookStatus(
+//        AnsibleRunnerHttpClient.PlaybookStatus playbookStatus,
+//        HttpResponse response
+//    ) throws IOException {
+//        when(httpClient.execute(any(HttpGet.class))).thenReturn(response);
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:106]: Why is the above code commented out? If non needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 12]
```diff
@@ -52,190 +34,190 @@
...
+//        when(httpClient.execute(any(HttpGet.class))).thenReturn(
+//            createHttpResponse("{\"status\": \"NOTFOUND\", \"msg\": \"not exists\"}", HttpStatus.SC_NOT_FOUND)
+//        );
 
-        assertThrows(AnsibleRunnerCallException.class, () -> {
-            client.getTotalEvents("123");
-        });
+//        assertThrows(AnsibleRunnerCallException.class, () -> {
+//            client.getTotalEvents("123");
+//        });
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:128]: Why is the above code commented out? If non needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 13]
```diff
@@ -52,190 +34,190 @@
...
+//                    "},",
+//                    "\"total_events\": 1",
+//                    "}",
+//                    "}"
+//                ),
+//                HttpStatus.SC_OK
+//            )
+//        );
+
+//        assertThat(client.processEvents("123", 0, null, null, Path.of("")), is(1));
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:152]: Why is the above code commented out? If non needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 14]
```diff
@@ -52,190 +34,190 @@
...
-        assertThrows(AnsibleRunnerCallException.class, () -> {
-            client.processEvents("123", 0, null, null, Path.of(""));
-        });
+//        when(httpClient.execute(any(HttpGet.class))).thenReturn(
+//            createHttpResponse("{\"status\": \"NOTFOUND\", \"msg\": \"not exists\"}", HttpStatus.SC_NOT_FOUND)
+//        );
+//
+//        assertThrows(AnsibleRunnerCallException.class, () -> {
+//            client.processEvents("123", 0, null, null, Path.of(""));
+//        });
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:163]: Why is the above code commented out? If non needed, please remove

File: backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java [hunk 15]
```diff
@@ -52,190 +34,190 @@
...
+//                    "}",
+//                    "}"
+//                ),
+//                HttpStatus.SC_OK
+//            )
+//        );
+//        assertThat(
+//            client.getYumPackages(any(String.class)), //TODO - fix test.
+//            containsInAnyOrder("mypackage1", "mypackage2")
+//        );
```
Comment (mwperina) [inline backend/manager/modules/bll/src/test/java/org/ovirt/engine/core/common/utils/ansible/AnsibleRunnerHttpClientTest.java:192]: Why is the above code commented out? If non needed, please remove

**Review by mwperina — CHANGES_REQUESTED** [review]

**Comment by michalskrivanek** [comment]
> when you're adding dependencies you cannot run it without those repos. that needs to be started manually

e.g. like in run # 32223, but you still need to add whatever else is missing (dwh, whatever) so that all requires ansible 2.12

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleCommandConfig.java (hunk 1 above)
Comment (dangel101) [reply backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/ansible/AnsibleCommandConfig.java:281]: absolutePath returns a path while we still need a String type to add to List<String> ansibleCommand

**Comment by mwperina** [comment]
Closing in favor of https://github.com/oVirt/ovirt-engine/pull/214

**Pull request closed without merging**